### PR TITLE
Implement CornerRadius for Cairo

### PR DESCRIPTION
**Pull request by ncarrillo**

### CODE REVIEW

**Comment by nc4rrillo** [comment]
Fixes #349

**Comment by jkoritzinsky** [comment]
Code looks good to me! I also notice that we don't have a render tests for corner radius on any of our rendering systems. Any chance you can add a render test for this?

I can help you with making the render test if you want.

**Comment by nc4rrillo** [comment]
I've never made one, I'd love to learn though.

**Comment by jkoritzinsky** [comment]
So you'll probably want to add your tests to the https://github.com/AvaloniaUI/Avalonia/blob/master/tests/Avalonia.RenderTests/Shapes/RectangleTests.cs file. Running the tests creates a *.out.png file (in the RenderToFile method) and compares it to the *.expected.png file with the same name (in the CompareImages method). The test file directories for the rectangle tests are ./tests/TestFiles/{RenderingSystem}/Shapes/Rectangle.

We don't have a good way to create the expected images, so we just run the test and generate the .out.png images. If the out images look right, then we rename them to be .expected.png. One way to do it is to run the test on one platform (like Direct2D1) and then copy the image into the other directories.

**Comment by danwalmsley** [comment]
@jkoritzinsky will try and add a render test for this, would be nice to get this merged in.

File: src/Gtk/Avalonia.Cairo/Media/DrawingContext.cs [hunk 1]
```diff
@@ -179,11 +179,25 @@ public void DrawGeometry(IBrush brush, Pen pen, Geometry geometry)
         /// <param name="rect">The rectangle bounds.</param>
         public void DrawRectangle(Pen pen, Rect rect, float cornerRadius)
         {
-            using (var p = SetPen(pen, rect.Size)) 
-            {
-                _context.Rectangle(rect.ToCairo ());
-                _context.Stroke();
-            }
+			if (cornerRadius == 0) {
+				using (var p = SetPen (pen, rect.Size)) {
```
Comment (mageos) [inline src/Gtk/Avalonia.Cairo/Media/DrawingContext.cs]: You can move the using statement to be outside the if\else block to reduce the duplication

File: src/Gtk/Avalonia.Cairo/Media/DrawingContext.cs [hunk 2]
```diff
@@ -179,11 +179,25 @@ public void DrawGeometry(IBrush brush, Pen pen, Geometry geometry)
...
+					_context.Rectangle (rect.ToCairo ());
+					_context.Stroke ();
+				}
+			} else {
+				using (var p = SetPen (pen, rect.Size)) {
+					double degrees = Math.PI / 180.0;
+					var cairoRect = rect.ToCairo ();
+
+					_context.NewSubPath ();
```
Comment (mageos) [inline src/Gtk/Avalonia.Cairo/Media/DrawingContext.cs]: This block for drawing a rectangle could be extracted into a method so that it is can be used in both DrawRectangle and FillRectangle

**Review by mageos — CHANGES_REQUESTED** [review]
Overall it looks pretty good, but I think there are some minor refactors that would make the code a bit cleaner and more maintainable going forward.

**Comment by grokys** [comment]
Closing this as it's very old and we're going to 🔥 cairo soon anyway. Thanks anyway @nc4rrillo !
